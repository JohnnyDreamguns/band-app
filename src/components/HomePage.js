import React from 'react';
import { Link } from 'react-router-dom';

const HomePage = ({ bands, loading, page, numOfPages, actions }) => (
  <div>
    {loading && <div className="loader"></div>}

    {bands &&
      bands.map(band => (
        <p key={band.id}>
          <Link to={`/band/${band.id}`}>{band.name}</Link>
        </p>
      ))}

    <button onClick={actions.setPreviousPage(page)} disabled={page === 1}>
      Previous Page
    </button>
    <button onClick={actions.setNextPage(page)} disabled={page === numOfPages}>
      Next Page
    </button>
  </div>
);

export default React.memo(HomePage);
