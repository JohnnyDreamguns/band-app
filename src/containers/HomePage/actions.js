import {
  SET_BAND_IDS,
  SET_PAGE_ID,
  SET_LOADING,
  CLEAR_HOME_PAGE
} from './constants';
import { doRequest } from '../../services/api';
import { setBands, setTotalNumBands } from '../../actions/data';
import { setError } from '../../actions/messages';
import { catchErrors } from '../../utils/catchErrors';

export const loadBands = payload => {
  return async dispatch => {
    dispatch(setLoading(true));

    const { data, error } = await catchErrors(
      doRequest({
        endpoint: 'getBandsByPage',
        params: { page: payload }
      })
    );

    if (error) {
      dispatch(setError(error));
    } else {
      dispatch(setBands(data.results));
      dispatch(setTotalNumBands(data.totalResults));
      dispatch(setBandIds(Object.keys(data.results).map(x => +x)));
    }
    dispatch(setLoading(false));
  };
};

export const setBandIds = payload => ({
  type: SET_BAND_IDS,
  payload
});

export const setPageId = payload => ({
  type: SET_PAGE_ID,
  payload
});

export const setLoading = payload => ({
  type: SET_LOADING,
  payload
});

export const clearHomepage = () => ({
  type: CLEAR_HOME_PAGE
});
