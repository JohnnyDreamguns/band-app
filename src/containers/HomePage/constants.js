export const LOAD_BANDS = 'bands-app/HomePage/LOAD_BANDS';
export const SET_BAND_IDS = 'bands-app/HomePage/SET_BAND_IDS';
export const SET_PAGE_ID = 'bands-app/HomePage/SET_PAGE_ID';
export const SET_LOADING = 'bands-app/HomePage/SET_LOADING';
export const CLEAR_HOME_PAGE = 'bands-app/HomePage/CLEAR_HOME_PAGE';
