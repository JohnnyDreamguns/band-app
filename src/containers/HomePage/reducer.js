import produce from 'immer';
import {
  SET_BAND_IDS,
  SET_PAGE_ID,
  SET_LOADING,
  CLEAR_HOME_PAGE
} from './constants';

// The initial state of the App
export const initialState = {
  bands: [],
  loading: false,
  page: 1
};

/* eslint-disable default-case, no-param-reassign */
const homeReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case SET_BAND_IDS:
        draft.bands = action.payload;
        break;

      case SET_PAGE_ID:
        draft.page = action.payload;
        break;

      case SET_LOADING:
        draft.loading = action.payload;
        break;

      case CLEAR_HOME_PAGE:
        draft.bands = [];
        break;
    }
  });

export default homeReducer;
