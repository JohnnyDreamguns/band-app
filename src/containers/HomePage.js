// React
import React from 'react';
import HomePage from '../components/HomePage';
import { useHomeData } from '../hooks/useHomeData';

const HomePageContainer = () => {
  const {
    bands,
    page,
    loading,
    numOfPages,
    useFetchHomePageData,
    setNextPage,
    setPreviousPage
  } = useHomeData();

  useFetchHomePageData(page);

  const componentProps = {
    bands,
    page,
    loading,
    numOfPages,
    actions: {
      setNextPage,
      setPreviousPage
    }
  };

  return <HomePage {...componentProps} />;
};

export default HomePageContainer;
