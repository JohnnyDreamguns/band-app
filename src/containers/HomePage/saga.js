import { call, put, takeLatest } from 'redux-saga/effects';
import { LOAD_BANDS } from './constants';
import { setBandIds, setLoading } from './actions';
import { setError } from '../../actions/messages';
import { setBands, setTotalNumBands } from '../../actions/data';

import { request } from '../../utils/request';

export function* getBandsSaga(action) {
  const requestURL = `http://localhost:3001/bands`;

  try {
    yield put(setLoading(true));
    const bands = yield call(request, `${requestURL}?page=${action.payload}`);

    yield put(setBands(bands.results));
    yield put(setTotalNumBands(bands.totalResults));
    yield put(setBandIds(Object.keys(bands.results).map(x => +x)));
    yield put(setLoading(false));
  } catch (err) {
    yield put(setLoading(false));
    yield put(setError('There was a problem loading the home data'));
  }
}

export default function* homepageDataSaga() {
  yield takeLatest(LOAD_BANDS, getBandsSaga);
}
