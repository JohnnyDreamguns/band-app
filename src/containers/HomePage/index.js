import React, { useEffect } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { loadBands, setPageId, clearHomepage } from './actions';
import { useInjectReducer } from '../../utils/injectReducer';
import {
  selectBands,
  selectLoading,
  selectPage,
  selectNumOfPages
} from './selectors';
import reducer from './reducer';

const key = 'homePage';

export const HomePage = ({
  loadBands,
  bands,
  loading,
  page,
  numOfPages,
  setPreviousPage,
  setNextPage,
  clearHomepage
}) => {
  useInjectReducer({ key, reducer });

  useEffect(() => {
    loadBands(page);

    return () => {
      clearHomepage();
    };
  }, [loadBands, clearHomepage, page]);

  return (
    <div>
      {loading && <div className="loader"></div>}

      {bands.map(band => (
        <p key={band.id}>
          <Link to={`/band/${band.id}`}>{band.name}</Link>
        </p>
      ))}

      <button onClick={setPreviousPage(page)} disabled={page === 1}>
        Previous Page
      </button>
      <button onClick={setNextPage(page)} disabled={page === numOfPages}>
        Next Page
      </button>
    </div>
  );
};

export const mapStateToProps = state => {
  return {
    bands: selectBands(state),
    loading: selectLoading(state),
    page: selectPage(state),
    numOfPages: selectNumOfPages(state)
  };
};

export const mapDispatchToProps = dispatch => ({
  loadBands: page => dispatch(loadBands(page)),
  setPreviousPage: page => () => dispatch(setPageId(page - 1)),
  setNextPage: page => () => dispatch(setPageId(page + 1)),
  clearHomepage: () => dispatch(clearHomepage())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(HomePage);
