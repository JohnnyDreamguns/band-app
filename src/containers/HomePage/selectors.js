import { initialState } from './reducer';
import { selectBandsData } from '../../selectors/data';

export const homeRoot = state => state.homePage || initialState;

export const selectPage = state => homeRoot(state)['page'];

export const selectBandIds = state => homeRoot(state)['bands'];

export const selectBands = state => {
  const bandIds = selectBandIds(state);
  if (!bandIds) return [];
  return bandIds
    .filter(x => selectBandsData(state)[x])
    .map(bandId => selectBandsData(state)[`${bandId}`]);
};

export const selectLoading = state => homeRoot(state)['loading'];

export const selectTotalNumBands = state => state.data.totalNumBands;

export const selectNumOfPages = state => {
  const numBands = selectTotalNumBands(state);
  if (numBands) {
    return Math.ceil(numBands / 3);
  }
};
