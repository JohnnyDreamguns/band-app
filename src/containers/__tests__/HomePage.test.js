import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import TestRenderer from 'react-test-renderer';
import HomePageContainer from '../HomePage';
import HomePage from '../../components/HomePage';

const mockUseFetchHomePageData = jest.fn();

jest.mock('../../hooks/useHomeData', () => ({
  useHomeData: () => ({
    bandIds: 'mock bandIds',
    bands: [{ id: 3, name: 'mock band' }],
    page: 2,
    loading: true,
    numOfPages: 3,
    setLoading: 'mock setLoading',
    setBandKeys: 'mock setBandKeys',
    setPage: 'mock setPage',
    fetchHomePageData: 'mock fetchHomePageData',
    useFetchHomePageData: mockUseFetchHomePageData,
    setNextPage: () => 'mock setNextPage',
    setPreviousPage: () => 'mock setPreviousPage'
  })
}));

jest.mock('../../hooks/useGlobalData', () => ({
  useGlobalData: () => ({
    setGlobalAlbums: () => {},
    setGlobalBands: () => {}
  })
}));

describe('HomePage container', () => {
  const testRenderer = TestRenderer.create(
    <BrowserRouter>
      <HomePageContainer />
    </BrowserRouter>
  );
  const testInstance = testRenderer.root;

  it('should match the props sent to the component', () => {
    expect(testInstance.findByType(HomePage.type).props).toMatchSnapshot();
  });

  it('should call mockUseFetchAlbumPageData', () => {
    expect(mockUseFetchHomePageData).toHaveBeenCalledTimes(1);
  });
});
