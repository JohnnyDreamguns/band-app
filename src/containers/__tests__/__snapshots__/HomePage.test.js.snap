// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage container should match the props sent to the component 1`] = `
Object {
  "actions": Object {
    "setNextPage": [Function],
    "setPreviousPage": [Function],
  },
  "bands": Object {
    "map": [Function],
  },
  "loading": true,
  "numOfPages": 3,
  "page": 2,
}
`;
